cpp/src/arrow/array/builder_base.cc

```diff
@@ -206,6 +207,45 @@ struct AppendScalarImpl {
     return Status::OK();
   }
 
+  Status Visit(const SparseUnionType& type) { return MakeUnionArray(type); }
+
+  Status Visit(const DenseUnionType& type) { return MakeUnionArray(type); }
+
+  template <typename T>
+  Status MakeUnionArray(const T& type) {
+    using BuilderType = typename TypeTraits<T>::BuilderType;
+
+    auto* builder = internal::checked_cast<BuilderType*>(builder_);
+    const auto count = n_repeats_ * (scalars_end_ - scalars_begin_);
+
+    RETURN_NOT_OK(builder->Reserve(count));
+
+    DCHECK_EQ(type.num_fields(), builder->num_children());
+    for (int field_index = 0; field_index < type.num_fields(); ++field_index) {
+      RETURN_NOT_OK(builder->child_builder(field_index)->Reserve(count));
+    }
+
+    for (int64_t i = 0; i < n_repeats_; i++) {
+      for (const std::shared_ptr<Scalar>* s = scalars_begin_; s != scalars_end_; s++) {
+        // For each scalar, 1. append the type code, 2. append the value to
+        // the corresponding child (and append null to the other children)
```

Comment:
This seems inappropriate for the dense array; for that case shouldn't we *only* append to the child builder which corresponds to the appended scalar's type code? What you've written doesn't produce an invalid dense union array but it will be somewhat pessimized since there will be unnecessary null slots

Reply:
Oh, you're right, indeed. Thanks for spotting this.